File: frontend/api/salesTracking.ts
import { apiClient } from "./client";
import type { MediaFormat } from "./types";

export type SalesProvider = "itad" | "platprices";

export interface IgnoredSalesTitle {
  provider: SalesProvider;
  gameId: number;
  gameUuid: string;
  gameName: string;
  gameSlug: string | null;
  gameCoverUrl: string | null;
  checkedAt: string | null;
}

export async function getIgnoredSalesTitles(): Promise<IgnoredSalesTitle[]> {
  const response = await apiClient.get<IgnoredSalesTitle[]>("/api/sales-tracking/ignored");
  return response.data;
}

export async function retryIgnoredSalesTitle(provider: SalesProvider, gameId: number): Promise<void> {
  await apiClient.post(`/api/sales-tracking/ignored/${provider}/${gameId}/retry`);
}

export async function removeIgnoredSalesTitle(provider: SalesProvider, gameId: number): Promise<void> {
  await apiClient.post(`/api/sales-tracking/ignored/${provider}/${gameId}/remove`);
}

export interface TrackedSalesItem {
  libraryItemId: number;
  gameId: number;
  gameUuid: string;
  gameName: string;
  gameSlug: string | null;
  gameCoverUrl: string | null;
  platformId: number | null;
  platformName: string | null;
  format: MediaFormat | null;
  digitalStorefront: string | null;
  targetPrice: number | null;
}

export async function getTrackedSalesItems(): Promise<TrackedSalesItem[]> {
  const response = await apiClient.get<TrackedSalesItem[]>("/api/sales-tracking/tracked");
  return response.data;
}

export async function untrackSalesItem(libraryItemId: number): Promise<void> {
  await apiClient.post(`/api/sales-tracking/tracked/${libraryItemId}/untrack`);
}

File: backend/app/api/routes/sales_tracking.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.schemas.sales_tracking import (
    IgnoredSalesTitleResponse,
    SalesProvider,
    TrackedSalesItemResponse,
    tracked_sales_item_from_orm,
)
from app.services import itad_service, library_service, platprices_service

router = APIRouter(prefix="/api/sales-tracking", tags=["sales-tracking"], dependencies=[Depends(get_current_user)])

_SERVICE_BY_PROVIDER = {"itad": itad_service, "platprices": platprices_service}


@router.get("/ignored", response_model=list[IgnoredSalesTitleResponse])
def list_ignored_sales_titles(db: Session = Depends(get_db)) -> list[IgnoredSalesTitleResponse]:
    items = [
        (item, "itad")
        for item in itad_service.list_ignored_items(db)
    ] + [
        (item, "platprices")
        for item in platprices_service.list_ignored_items(db)
    ]
    responses = [
        IgnoredSalesTitleResponse(
            provider=provider,
            game_id=item.game.id,
            game_uuid=item.game.uuid,
            game_name=item.game.name,
            game_slug=item.game.slug,
            game_cover_url=item.game.cover_url,
            checked_at=item.checked_at,
        )
        for item, provider in items
    ]
    responses.sort(key=lambda entry: entry.game_name.lower())
    return responses


@router.post("/ignored/{provider}/{game_id}/retry", status_code=status.HTTP_204_NO_CONTENT)
def retry_ignored_sales_title(provider: SalesProvider, game_id: int, db: Session = Depends(get_db)) -> None:
    _SERVICE_BY_PROVIDER[provider].retry_ignored_item(db, game_id)


@router.post("/ignored/{provider}/{game_id}/remove", status_code=status.HTTP_204_NO_CONTENT)
def remove_ignored_sales_title(provider: SalesProvider, game_id: int, db: Session = Depends(get_db)) -> None:
    _SERVICE_BY_PROVIDER[provider].remove_ignored_item(db, game_id)


@router.get("/tracked", response_model=list[TrackedSalesItemResponse])
def list_tracked_sales_items(db: Session = Depends(get_db)) -> list[TrackedSalesItemResponse]:
    items = library_service.list_tracked_items(db)
    return [tracked_sales_item_from_orm(item) for item in items]


@router.post("/tracked/{library_item_id}/untrack", status_code=status.HTTP_204_NO_CONTENT)
def untrack_sales_item(library_item_id: int, db: Session = Depends(get_db)) -> None:
    library_service.update_library_item(db, library_item_id, track_for_sales=False)
